README.md
#django-ldap3-sync
django-ldap3-sync is a fork of [django-ldap-sync](https://github.com/jbittel/django-ldap-sync) originally created by Jason Bittel.
django-ldap3-sync introduces the following features:

 - Uses the [ldap3 library](https://github.com/cannatag/ldap3) for ldap communication. ldap3 is pure python and python 3 compatible.
 - Can synchronize group membership directly out of the LDAP directory.
 - Can manage deletion of groups / users in the directory by suspending or deleting those objects in Django.
 - Will update existing Django users groups if information changes in the directory.
 - Paged LDAP searches
 - LDAP Server Pools

django-ldap3-sync provides a Django management command that synchronizes LDAP
users and groups from an authoritative server. It performs a one-way
synchronization that creates and/or updates the local Django users and groups. It can also optionally delete or suspend users or groups that are no longer present in the directory.

This synchronization is performed each time the management command is run and
can be fired manually on demand, via an automatic cron script or as a periodic
`Celery` task.

##Quickstart
1. Install the application::

      `pip install django-ldap3-sync`

2. Append it to the installed apps::

      ```
      INSTALLED_APPS = (
          # ...
          'ldap3_sync',
      )
      ```

3. Configure the required `settings`.

4. Run the synchronization management command::

      `manage.py syncldap`

For more information on configuration see the documentation below.


----------

##Configuration
**User Synchronization Configuration Items**

***LDAP_SYNC_USER_FILTER***  
Default: `(objectClass=user)`  
The filter used to retrieve users from the directory. Must be in standard LDAP filter syntax as per [RFC2254](http://www.ietf.org/rfc/rfc2254.txt?number=2254)

***LDAP_SYNC_USER_BASE***  
Default: The value of `LDAP_SYNC_BASE`  
The distinguished name of the container to base the search for users in.

***LDAP_SYNC_USER_ATTRIBUTES***  
Default: **No Default -- Required Item**  
A dictionary of key value pairs where the keys are the names of ldap fields and the values are the names of corresponding django model fields. New users will be created with these fields populated and existing users will have these fields updated. *This dictionary must map the Django User model username field to an LDAP object field.*

***LDAP_SYNC_USER_EXEMPT_FROM_SYNC***  
Default: `[]`  
A list of usernames corresponding to Django users who should be excluded from the sync. Useful for Administrative users who do not have a corresponding user in the directory.

***LDAP_SYNC_USER_REMOVAL_ACTION***  
Default: `NOTHING`  
The action to take when a user no longer exists in the directory. Possible values are `NOTHING`, `SUSPEND` and `DELETE`. Note that the `SUSPEND` option uses the Django user models is_active field and sets it to False.

***LDAP_SYNC_USER_SET_UNUSABLE_PASSWORD***  
Default: `True`  
If true this uses the django method set_unusable_password on all newly created users. Useful where django authentication will not be used.

***LDAP_SYNC_USERS***  
Default: `True`  
Controls whether users should be synchronized from the directory.

**Group Synchronization Configuration Items**  

***LDAP_SYNC_GROUP_FILTER***  
Default: `(objectClass=group)`  
The filter used to retrieve groups from the directory. Must be in standard LDAP filter syntax as per [RFC2254](http://www.ietf.org/rfc/rfc2254.txt?number=2254)

***LDAP_SYNC_GROUP_BASE***  
Default: The value of `LDAP_SYNC_BASE`  
The distinguished name of the container to base the search for groups in.

***LDAP_SYNC_GROUP_ATTRIBUTES***  
Default: **No Default -- Required Item**  
A dictionary of key value pairs where the keys are the names of LDAP fields and the values are the names of corresponding Django model fields. New groups will be created with these fields populated and existing users will have these fields updated.

***LDAP_SYNC_GROUP_REMOVAL_ACTION***  
Default: `NOTHING`  
The action to take when a group no longer exists in the directory. Possible values are `NOTHING` and `DELETE`.

***LDAP_SYNC_GROUP_EXEMPT_FROM_SYNC***  
Default: `[]`  
A list of group names of Django groups that should be excluded from the sync.

***LDAP_SYNC_GROUPS***  
Default: `True`  
Controls wether groups should be synchronized.

**Membership Synchronization Configuration Items**  

***LDAP_SYNC_GROUP_MEMBERSHIP***  
Default: `True`  
Controls wether groups will be synchronized from the directory.

***LDAP_SYNC_GROUP_MEMBERSHIP_FILTER***  
Default: `(&(objectClass=group)(member={user_dn}))`  
The filter used to retrieve the groups that a user belongs to. `{user_dn}` will be replaced with the distinguished name of the user.

**LDAP Server Options**  

***LDAP_CONFIG***  
Default: **No Default -- Required Item**  
Configuration item used to configure the server pool. `LDAP_CONFIG` can contain the following keys:

 - **page_size**  
	Default: `500`  
	The page size for searches using this server pool.  
 - **bind_user**  
    Default: `None`  
    The distinguished name of the user to bind to the directory with.  
 - **bind_password**  
   Default: `None`  
   The password of the user to bind to the directory with.  
 - **pooling_strategy**  
   Default: `ROUND_ROBIN`  
   The strategy to use when the pool contains multiple servers. See ldap3 documentation at      https://ldap3.readthedocs.org/en/latest/servers.html#server-pool for more information. Can be   `FIRST`, `ROUND_ROBIN` or `RANDOM`.  
 - **servers**  
Default: `None`  
A list of dictionaries each one containing configuration information for a server. Possible server configuration keys are:  
	 - **address**  
		Default: **No Default -- Required Item**  
		Either the IP address or FQDN of the directory server.  
	 - **use_ssl**  
		Default: `False`  
		Use SSL with this conntection.  
	 - **port**  
	   Default: `389`  
	   The port to use with this server.  
	 - **timeout**  
		Default: `30`  
		Connection timeout with this server.  
	 - **get_schema**  
		Default: `SCHEMA`  
	Determines which schema information to retrieve from the server. At a minimum this should be SCHEMA so that values retrieved from the directory are coerced to proper python types.  


docs/changelog.md
Changelog
=========

These are the notable changes for each django-ldap3-sync release. For
additional detail, read the complete [Commit History](https://github.com/pgower/django-ldap3-sync/commits/).

**django-ldap3-sync 0.1**
   * Fork django-ldap-sync
   * Add group membership synchronisation
   * Rewrite to use ldap3 as ldap library

**django-ldap-sync 0.1.1**
   * Fix exception with AD internal referrals

**django-ldap-sync 0.1.0**
   * Initial release


testing/testing/wsgi.py
"""
WSGI config for testing project.

It exposes the WSGI callable as a module-level variable named ``application``.

For more information on this file, see
https://docs.djangoproject.com/en/1.6/howto/deployment/wsgi/
"""

import os
os.environ.setdefault("DJANGO_SETTINGS_MODULE", "testing.settings")

from django.core.wsgi import get_wsgi_application
application = get_wsgi_application()


testing/testing/settings.py
"""
Django settings for testing project.

For more information on this file, see
https://docs.djangoproject.com/en/1.6/topics/settings/

For the full list of settings and their values, see
https://docs.djangoproject.com/en/1.6/ref/settings/
"""

# Build paths inside the project like this: os.path.join(BASE_DIR, ...)
import os
BASE_DIR = os.path.dirname(os.path.dirname(__file__))


# Quick-start development settings - unsuitable for production
# See https://docs.djangoproject.com/en/1.6/howto/deployment/checklist/

LOGGING = {
    'version': 1,
    'disable_existing_loggers': True,
    'formatters': {
        'verbose': {
            'format': '%(levelname)s %(asctime)s %(module)s %(process)d %(thread)d %(message)s'
        },
        'simple': {
            'format': '%(levelname)s %(message)s'
        },
    },
    'handlers': {
        'null': {
            'level': 'DEBUG',
            'class': 'django.utils.log.NullHandler',
        },
        'console': {
            'level': 'DEBUG',
            'class': 'logging.StreamHandler',
            'formatter': 'simple'
        },
        'mail_admins': {
            'level': 'ERROR',
            'class': 'django.utils.log.AdminEmailHandler',
        }
    },
    'loggers': {
        'django': {
            'handlers': ['null'],
            'propagate': True,
            'level': 'INFO',
        },
        'django.request': {
            'handlers': ['mail_admins'],
            'level': 'ERROR',
            'propagate': False,
        },
        'ldap_sync': {
            'handlers': ['console'],
            'propogate': True,
            'level': 'DEBUG',
        },
    }
}

# SECURITY WARNING: keep the secret key used in production secret!
SECRET_KEY = '^e!ya!zqt$_z$gyevd&g86o62i&3a_2@=gd%^1#e&rop2t9xq2'

# SECURITY WARNING: don't run with debug turned on in production!
DEBUG = True

TEMPLATE_DEBUG = True

ALLOWED_HOSTS = []


# Application definition

INSTALLED_APPS = (
    'django.contrib.admin',
    'django.contrib.auth',
    'django.contrib.contenttypes',
    'django.contrib.sessions',
    'django.contrib.messages',
    'django.contrib.staticfiles',
    'ldap_sync',
)

MIDDLEWARE_CLASSES = (
    'django.contrib.sessions.middleware.SessionMiddleware',
    'django.middleware.common.CommonMiddleware',
    'django.middleware.csrf.CsrfViewMiddleware',
    'django.contrib.auth.middleware.AuthenticationMiddleware',
    'django.contrib.messages.middleware.MessageMiddleware',
    'django.middleware.clickjacking.XFrameOptionsMiddleware',
)

ROOT_URLCONF = 'testing.urls'

WSGI_APPLICATION = 'testing.wsgi.application'


# Database
# https://docs.djangoproject.com/en/1.6/ref/settings/#databases

DATABASES = {
    'default': {
        'ENGINE': 'django.db.backends.sqlite3',
        'NAME': os.path.join(BASE_DIR, 'db.sqlite3'),
    }
}

# Internationalization
# https://docs.djangoproject.com/en/1.6/topics/i18n/

LANGUAGE_CODE = 'en-us'

TIME_ZONE = 'UTC'

USE_I18N = True

USE_L10N = True

USE_TZ = True


# Static files (CSS, JavaScript, Images)
# https://docs.djangoproject.com/en/1.6/howto/static-files/

STATIC_URL = '/static/'

# LDAP_SYNC Configuration
LDAP_CONFIG = {
    'page_size': 100,
    'bind_user': 'LDAPBinder',
    'bind_password': '5tm0NICa5198SIx',
    'servers': [
        {'address': '10.192.81.55', 'port': 389, 'use_ssl': False, 'timeout': 30},
        {'address': '10.192.81.54', 'port': 389, 'use_ssl': False, 'timeout': 30},
    ]

}

# LDAP SYNC SETTINGS
LDAP_SYNC_BASE = 'OU=Automated Objects,DC=stmonicas,DC=qld,DC=edu,DC=au'
LDAP_SYNC_USER_FILTER = '(&(objectClass=user)(|(employeeType=STUDENT)(employeeType=STAFF)))'
LDAP_SYNC_USER_ATTRIBUTES = {
    "sAMAccountName": "username",
    "givenName": "first_name",
    "sn": "last_name",
    "mail": "email",
}
LDAP_SYNC_GROUP_FILTER = "(&(objectClass=group)(|(sAMAccountName=*_STAFF_*)(sAMAccountName=*_STUDENT_*)))"
LDAP_SYNC_GROUP_ATTRIBUTES = {
    "sAMAccountName": "name",
}

ldap3_sync/models.py
from django.db import models
from django.contrib.auth.models import User, Group


# Store LDAP info about the created groups so that we can easily
# identify them in subsequent syncs

HELP_TEXT = ('DO NOT edit this unless you really know '
            'what your doing. It is much safer to delete '
            'this entire record and let the sync command '
            'recreate it.')


class LDAPUser(models.Model):
    obj = models.OneToOneField(User, related_name='ldap_sync_user', verbose_name='User')
    # There does not appear to be a maximum length for distinguishedName
    # safest to use text to avoid any length issues down the track
    distinguished_name = models.TextField(blank=True, help_text=HELP_TEXT)

    def __unicode__(self):
        return '{} {} ({})'.format(self.obj.first_name,
                                   self.obj.last_name,
                                   self.distinguished_name)

    class Meta:
        verbose_name = 'LDAP User'
        verbose_name_plural = 'LDAP Users'


class LDAPGroup(models.Model):
    obj = models.OneToOneField(Group, related_name='ldap_sync_group', verbose_name='Group')
    distinguished_name = models.TextField(blank=True, help_text=HELP_TEXT)

    def __unicode__(self):
        return '{} ({})'.format(self.obj.name, self.distinguished_name)

    class Meta:
        verbose_name = 'LDAP Group'
        verbose_name_plural = 'LDAP Groups'


ldap3_sync/management/commands/syncldap.py
import logging

import ldap3
from ldap3.core.exceptions import LDAPExceptionError, LDAPCommunicationError

from django.conf import settings
from django.core.management.base import NoArgsCommand
from django.contrib.auth import get_user_model
from django.contrib.auth.models import Group
from django.core.exceptions import ImproperlyConfigured
from django.db import IntegrityError, DataError
from ldap3_sync.models import LDAPUser, LDAPGroup


logger = logging.getLogger(__name__)

NOTHING = 'NOTHING'
SUSPEND = 'SUSPEND'
DELETE = 'DELETE'

USER_REMOVAL_OPTIONS = (NOTHING, SUSPEND, DELETE)
GROUP_REMOVAL_OPTIONS = (NOTHING, DELETE)


class Command(NoArgsCommand):
    help = "Synchronize users, groups and group membership from an LDAP server"

    def handle_noargs(self, **options):
        self.load_settings()
        self.sync_ldap_users()
        self.sync_ldap_groups()

    def get_ldap_users(self):
        """
        Retrieve user data from target LDAP server.
        """
        logging.debug('Retrieving Users from LDAP')
        users = self.smart_ldap_searcher.search(self.user_base, self.user_filter, ldap3.SEARCH_SCOPE_WHOLE_SUBTREE, self.user_ldap_attribute_names)
        logger.info("Retrieved {} LDAP users".format(len(users)))
        return users

    def sync_ldap_users(self):
        """
        Synchronize users with local user database.
        """
        ldap_users = self.get_ldap_users()
        django_users = self.get_django_users()

        self.sync_generic(ldap_objects=ldap_users,
                          django_objects=django_users,
                          attribute_map=self.user_attribute_map,
                          django_object_model=self.user_model,
                          unique_name_field='username',
                          ldap_sync_model=LDAPUser,
                          ldap_sync_related_name='ldap_sync_user',
                          exempt_unique_names=self.exempt_usernames,
                          removal_action=self.user_removal_action)

    def get_ldap_groups(self):
        """
        Retrieve groups from target LDAP server.
        """
        logger.debug('Retrieving Groups from LDAP')
        return self.smart_ldap_searcher.search(self.group_base, self.group_filter, ldap3.SEARCH_SCOPE_WHOLE_SUBTREE, self.group_ldap_attribute_names)

    def sync_ldap_groups(self):
        """
        Synchronize LDAP groups with local group database.
        """
        ldap_groups = self.get_ldap_groups()
        django_groups = self.get_django_groups()

        self.sync_generic(ldap_objects=ldap_groups,
                          django_objects=django_groups,
                          attribute_map=self.group_attribute_map,
                          django_object_model=Group,
                          unique_name_field='name',
                          ldap_sync_model=LDAPGroup,
                          ldap_sync_related_name='ldap_sync_group',
                          exempt_unique_names=self.exempt_groupnames,
                          removal_action=self.group_removal_action)

    def get_ldap_group_membership(self, user_dn):
        '''
        Retrieve a list of django groups that this user DN is a member of.
        '''
        if not hasattr(self, '_group_cache'):
            self._group_cache = {}
        logger.debug('Retrieving groups that {} is a member of'.format(user_dn))
        ldap_groups = self.smart_ldap_searcher.search(self.group_base, self.group_membership_filter.format(user_dn=user_dn), ldap3.SEARCH_SCOPE_SINGLE_LEVEL, None)
        django_groups = []
        for ldap_group in ldap_groups:
            group_dn = ldap_group['dn']
            if group_dn in self._group_cache:
                django_groups.append(self._group_cache[group_dn])
            else:
                try:
                    ldap_sync_group = LDAPGroup.objects.get(dn=group_dn)
                    django_groups.append(ldap_sync_group.obj)
                    self._group_cache[group_dn] = ldap_sync_group.obj
                except LDAPGroup.DoesNotExist:
                    logger.warning('Cannot find Django Group associated with DN {}'.format(ldap_group['dn']))
                    continue
        return django_groups

    def sync_group_membership(self):
        '''
        Synchornize group membership with the directory. Only synchronize groups that have a related LDAPGroup object.
        '''
        django_users = self.get_django_users()
        for django_user in django_users:
            try:
                user_dn = django_user.ldap_sync_user.dn
            except LDAPUser.DoesNotExist:
                logger.warning('Django user with {} = {} does not have a distinguishedName associated'.format(self.username_field, getattr(django_user, self.username_field)))
                continue
            django_groups = self.get_ldap_group_membership(user_dn)
            django_user.groups = django_groups
            django_user.save()

    def sync_generic(self,
                     ldap_objects,
                     django_objects,
                     attribute_map,
                     django_object_model,
                     unique_name_field,
                     ldap_sync_model,
                     ldap_sync_related_name,
                     exempt_unique_names=[],
                     removal_action=NOTHING):
        """
        A generic synchronization method.
        """
        model_name = django_object_model.__name__

        unsaved_models = []
        model_dn_map = {}

        updated_model_count = 0

        for ldap_object in ldap_objects:
            try:
                value_map = self.generate_value_map(attribute_map, ldap_object['attributes'])
            except MissingLdapField as e:
                logger.error('LDAP Object {} is missing a field: {}'.format(ldap_object['dn'], e))
                continue
            unique_name = value_map[unique_name_field]
            distinguished_name = ldap_object['dn']

            model_dn_map[unique_name] = distinguished_name

            try:
                django_object = django_objects[unique_name]
                if self.will_model_change(value_map, django_object):
                    self.apply_value_map(value_map, django_object)
                    django_object.save()
                    updated_model_count += 1
                try:
                    if getattr(django_object, ldap_sync_related_name).distinguished_name != distinguished_name:
                        getattr(django_object, ldap_sync_related_name).distinguished_name = distinguished_name
                        getattr(django_object, ldap_sync_related_name).save()
                except ldap_sync_model.DoesNotExist:
                    ldap_sync_model(obj=django_object, distinguished_name=distinguished_name).save()
                del(django_objects[unique_name])
            except KeyError:
                django_object = django_object_model(**value_map)
                if hasattr(django_object, 'set_unusable_password') and self.set_unusable_password:
                    # only do this when its a user (or has this method) and the config says to do it
                    django_object.set_unusable_password()
                unsaved_models.append(django_object)
        logger.debug('Bulk creating unsaved {}'.format(model_name))
        self.chunked_bulk_create(django_object_model, unsaved_models)
        # django_object_model.objects.bulk_create(unsaved_models)
        logger.debug('Retrieving ID\'s for the objects that were just created')

        filter_key = '{}__in'.format(unique_name_field)
        filter_value = [getattr(u, unique_name_field) for u in unsaved_models]
        just_saved_models = django_object_model.objects.filter(**{filter_key: filter_value}).all()
        logger.debug('Bulk creating ldap_sync models')
        new_ldap_sync_models = [ldap_sync_model(obj=u, distinguished_name=model_dn_map[getattr(u, unique_name_field)]) for u in just_saved_models]
        self.chunked_bulk_create(ldap_sync_model, new_ldap_sync_models)

        msg = 'Updated {} existing {}'.format(updated_model_count, model_name)
        self.stdout.write(msg)
        logger.info(msg)

        msg = 'Created {} new {}'.format(len(unsaved_models), model_name)
        self.stdout.write(msg)
        logger.info(msg)

        # Anything left in the existing_users dict is no longer in the ldap directory
        # These should be disabled.
        existing_unique_names = set(_unique_name for _unique_name in django_objects.keys())
        # existing_unique_names.difference_update(exempt_unique_names)
        existing_model_ids = [djo.id for djo in django_objects.values() if getattr(djo, unique_name_field) in existing_unique_names]

        if removal_action == NOTHING:
            logger.info('Removal action is set to NOTHING so the {} objects that would have been removed are being ignored.'.format(len(existing_unique_names)))
        elif removal_action == SUSPEND:
            if hasattr(django_object_model, 'is_active'):
                django_object_model.objects.in_bulk(existing_model_ids).update(is_active=False)
                logger.info('Suspended {} {}.'.format(len(existing_model_ids), model_name))
            else:
                logger.info('REMOVAL_ACTION is set to SUSPEND however {} do not have an is_active attribute. Effective action will be NOTHING for {}.'.format(model_name, len(existing_model_ids)))
        elif removal_action == DELETE:
            django_object_model.objects.filter(id__in=existing_model_ids).all().delete()
            logger.info('Deleted {} {}.'.format(len(existing_unique_names), model_name))

        logger.info("{} are synchronized".format(model_name))
        self.stdout.write('{} are synchronized'.format(model_name))

    def will_model_change(self, value_map, user_model):
        # I think all the attrs are utf-8 strings, possibly need to coerce
        # local user values to strings?
        for model_attr, value in value_map.items():
            if not getattr(user_model, model_attr) == value:
                return True
        return False

    def chunked_bulk_create(self, django_model_object, unsaved_models, chunk_size=None):
        if chunk_size is None:
            chunk_size = self.bulk_create_chunk_size
        for i in range(0, len(unsaved_models), chunk_size):
            django_model_object.objects.bulk_create(unsaved_models[i:i + chunk_size])

    def apply_value_map(self, value_map, user_model):
        for k, v in value_map.items():
            try:
                setattr(user_model, k, v)
            except AttributeError:
                raise UnableToApplyValueMapError('User model {} does not have attribute {}'.format(user_model.__class__.__name__, k))
        return user_model

    def generate_value_map(self, attribute_map, ldap_attribute_values):
        '''Given an attribute map (dict with keys as ldap attrs and values as model attrs) generate a dictionary
           which maps model attribute keys to ldap values'''
        value_map = {}
        for ldap_attr, model_attr in attribute_map.items():
            try:
                value_map[model_attr] = ldap_attribute_values[ldap_attr]
            except KeyError:
                raise MissingLdapField(ldap_attr)
        return value_map

    def get_django_objects(self, model):
        '''
        Given a Django model class get all of the current records that match.
        This is better than django's bulk methods and has no upper limit.
        '''
        model_name = model.__class__.__name__
        model_objects = [i for i in model.objects.all()]
        logger.debug('Found {} {} objects in DB'.format(len(model_objects), model_name))
        return model_objects

    def get_django_users(self):
        '''
        Return a dictionary of all existing users where the key is the username and the value is the user object.
        '''
        return dict([(getattr(u, self.username_field), u) for u in self.get_django_objects(self.user_model) if getattr(u, self.username_field) not in self.exempt_usernames])

    def get_django_groups(self):
        '''
        Return a dictionary of all existing groups where the key is the group name and the value is the group object.
        DO NOT return any groups whose name in in the LDAP_SYNC_GROUP_EXEMPT_FROM_SYNC collection.
        '''
        return dict([(g.name, g) for g in self.get_django_objects(Group) if g.name not in self.exempt_groupnames])

    def load_settings(self):
        '''
        Get all of the required settings to perform a sync and check them for sanity.
        '''
        self.bulk_create_chunk_size = getattr(settings, 'LDAP_SYNC_BULK_CREATE_CHECK_SIZE', 50)

        # User sync settings
        self.user_filter = getattr(settings, 'LDAP_SYNC_USER_FILTER', '(objectClass=user)')

        try:
            self.user_base = getattr(settings, 'LDAP_SYNC_USER_BASE')
        except AttributeError:
            try:
                self.user_base = getattr(settings, 'LDAP_SYNC_BASE')
            except AttributeError:
                raise ImproperlyConfigured('Either LDAP_SYNC_USER_BASE or LDAP_SYNC_BASE are required. Neither were found.')

        try:
            self.user_attribute_map = getattr(settings, 'LDAP_SYNC_USER_ATTRIBUTES')
        except AttributeError:
            raise ImproperlyConfigured('LDAP_SYNC_USER_ATTRIBUTES is a required setting')
        self.user_ldap_attribute_names = self.user_attribute_map.keys()
        self.user_model_attribute_names = self.user_attribute_map.values()

        self.exempt_usernames = getattr(settings, 'LDAP_SYNC_USER_EXEMPT_FROM_SYNC', [])
        self.user_removal_action = getattr(settings, 'LDAP_SYNC_USER_REMOVAL_ACTION', NOTHING)
        if self.user_removal_action not in USER_REMOVAL_OPTIONS:
            raise ImproperlyConfigured('LDAP_SYNC_USER_REMOVAL_ACTION must be one of {}'.format(USER_REMOVAL_OPTIONS))

        self.user_model = get_user_model()
        self.username_field = getattr(self.user_model, 'USERNAME_FIELD', 'username')

        self.set_unusable_password = getattr(settings, 'LDAP_SYNC_USER_SET_UNUSABLE_PASSWORD', True)

        self.sync_users = getattr(settings, 'LDAP_SYNC_USERS', True)

        # Check to make sure we have assigned a value to the username field
        if self.username_field not in self.user_model_attribute_names:
            raise ImproperlyConfigured("LDAP_SYNC_USER_ATTRIBUTES must contain the username field '%s'" % self.username_field)

        # Group sync settings
        self.group_filter = getattr(settings, 'LDAP_SYNC_GROUP_FILTER', '(objectClass=group)')

        try:
            self.group_base = getattr(settings, 'LDAP_SYNC_GROUP_BASE')
        except AttributeError:
            try:
                self.group_base = getattr(settings, 'LDAP_SYNC_BASE')
            except AttributeError:
                    raise ImproperlyConfigured('Either LDAP_SYNC_GROUP_BASE or LDAP_SYNC_BASE are required. Neither were found.')

        try:
            self.group_attribute_map = getattr(settings, 'LDAP_SYNC_GROUP_ATTRIBUTES')
        except AttributeError:
            raise ImproperlyConfigured('LDAP_SYNC_GROUP_ATTRIBUTES is a required setting')
        self.group_ldap_attribute_names = self.group_attribute_map.keys()
        self.group_model_attribute_names = self.group_attribute_map.values()

        self.group_removal_action = getattr(settings, 'LDAP_SYNC_GROUP_REMOVAL_ACTION', NOTHING)
        if self.group_removal_action not in GROUP_REMOVAL_OPTIONS:
            raise ImproperlyConfigured('LDAP_SYNC_GROUP_REMOVAL_ACTION must be one of {}'.format(GROUP_REMOVAL_OPTIONS))

        self.exempt_groupnames = getattr(settings, 'LDAP_SYNC_GROUP_EXEMPT_FROM_SYNC', [])

        self.sync_groups = getattr(settings, 'LDAP_SYNC_GROUPS', True)

        self.sync_group_membership = getattr(settings, 'LDAP_SYNC_GROUP_MEMBERSHIP', True)

        self.group_membership_filter = getattr(settings, 'LDAP_SYNC_GROUP_MEMBERSHIP_FILTER', '(&(objectClass=group)(member={user_dn}))')


        # LDAP Servers
        try:
            self.ldap_config = getattr(settings, 'LDAP_CONFIG')
        except AttributeError:
            raise ImproperlyConfigured('LDAP_CONFIG is a required configuration item')
        self.smart_ldap_searcher = SmartLDAPSearcher(self.ldap_config)


class SmartLDAPSearcher:
    def __init__(self, ldap_config):
        self.ldap_config = ldap_config
        # Setup a few other config items
        self.page_size = self.ldap_config.get('page_size', 500)
        self.bind_user = self.ldap_config.get('bind_user', None)
        self.bind_password = self.ldap_config.get('bind_password', None)
        pooling_strategy = self.ldap_config.get('pooling_strategy', 'ROUND_ROBIN')
        if pooling_strategy not in ldap3.POOLING_STRATEGIES:
            raise ImproperlyConfigured('LDAP_CONFIG.pooling_strategy must be one of {}'.format(ldap3.POOLING_STRATEGIES))
        self.server_pool = ldap3.ServerPool(None, pooling_strategy)
        logger.debug('Created new LDAP Server Pool with pooling strategy: {}'.format(pooling_strategy))
        try:
            server_defns = self.ldap_config.get('servers')
        except AttributeError:
            raise ImproperlyConfigured('ldap_config.servers must be defined and must contain at least one server')
        for server_defn in server_defns:
            self.server_pool.add(self._defn_to_server(server_defn))

    def _defn_to_server(self, defn):
        '''Turn a settings file server definition into a ldap3 server object'''
        try:
            address = defn.get('address')
        except AttributeError:
            raise ImproperlyConfigured('Server definition must contain an address')
        port = defn.get('port', 389)
        use_ssl = defn.get('use_ssl', False)
        timeout = defn.get('timeout', 30)
        get_info = defn.get('get_schema', ldap3.SCHEMA)
        return ldap3.Server(address, port=port, use_ssl=use_ssl, connect_timeout=timeout, get_info=get_info)

    def get_connection(self):
        c = ldap3.Connection(self.server_pool, user=self.bind_user, password=self.bind_password)
        r = c.bind()
        logger.debug('Trying to bind connection, result is: {}'.format(r))
        return c

    def search(self, base, filter, scope, attributes):
        '''Perform a paged search but return all of the results in one hit'''
        logger.debug('SmartLDAPSearcher.search called with base={}, filter={}, scope={} and attributes={}'.format(str(base), str(filter), str(scope), str(attributes)))
        connection = self.get_connection()
        connection.search(search_base=base, search_filter=filter, search_scope=scope, attributes=attributes, paged_size=self.page_size, paged_cookie=None)
        logger.debug('Connection.search.response is: {}'.format(connection.response))
        if len(connection.response) < self.page_size:
            results = connection.response
        else:
            results = connection.response
            cookie = connection.result['controls']['1.2.840.113556.1.4.319']['value']['cookie']
            while cookie:
                connection.search(search_base=base, search_filter=filter, search_scope=ldap3.SEARCH_SCOPE_WHOLE_SUBTREE, attributes=attributes, paged_size=self.page_size, paged_cookie=cookie)
                results += connection.response
                cookie = connection.result['controls']['1.2.840.113556.1.4.319']['value']['cookie']
        connection.unbind()
        return results

    def get(self, dn, attributes=[]):
        '''Return the object referenced by the given dn or return None'''
        # break the dn down and get a base from it
        search_base = ','.join(dn.split(',')[1:])
        connection = self.get_connection()
        connection.search(search_base=search_base, search_filter='(distinguishedName={})'.format(dn), search_scope=ldap3.SEARCH_SCOPE_SINGLE_LEVEL, attributes=attributes)
        results = connection.response
        if len(results) > 1:
            raise MultipleLDAPResultsReturned()
        elif len(results) == 0:
            return None
        else:
            return results[0]


class UnableToApplyValueMapError(Exception):
    pass


class MissingLdapField(Exception):
    pass


class SyncError(Exception):
    pass


class MultipleLDAPResultsReturned(Exception):
    pass

# class PagedResultsSearchObject:
#     """
#     Taken from the python-ldap paged_search_ext_s.py demo, showing how to use
#     the paged results control: https://bitbucket.org/jaraco/python-ldap/
#     """
#     page_size = getattr(settings, 'LDAP_SYNC_PAGE_SIZE', 100)

#     def paged_search_ext_s(self, base, scope, filterstr='(objectClass=*)',
#                            attrlist=None, attrsonly=0, serverctrls=None,
#                            clientctrls=None, timeout=-1, sizelimit=0):
#         """
#         Behaves exactly like LDAPObject.search_ext_s() but internally uses the
#         simple paged results control to retrieve search results in chunks.
#         """
#         req_ctrl = SimplePagedResultsControl(True, size=self.page_size,
#                                              cookie='')

#         # Send first search request
#         msgid = self.search_ext(base, ldap.SCOPE_SUBTREE, filterstr,
#                                 attrlist=attrlist,
#                                 serverctrls=(serverctrls or []) + [req_ctrl])
#         results = []

#         while True:
#             rtype, rdata, rmsgid, rctrls = self.result3(msgid)
#             results.extend(rdata)
#             # Extract the simple paged results response control
#             pctrls = [c for c in rctrls if c.controlType ==
#                       SimplePagedResultsControl.controlType]

#             if pctrls:
#                 if pctrls[0].cookie:
#                     # Copy cookie from response control to request control
#                     req_ctrl.cookie = pctrls[0].cookie
#                     msgid = self.search_ext(base, ldap.SCOPE_SUBTREE,
#                                             filterstr, attrlist=attrlist,
#                                             serverctrls=(serverctrls or []) +
#                                             [req_ctrl])
#                 else:
#                     break

#         return results


# class PagedLDAPObject(LDAPObject, PagedResultsSearchObject):
#     pass
